iris_grib/tests/unit/load_convert/test_product_definition_section.py

```diff
@@ -0,0 +1,134 @@
+class TestFixedSurfaces(Test):
+    def generic_fixed_surface_test(self, fs_is_expected, fs_is_present):
+        # Whether or not fixed surface elements are expected/present in the
+        # section 4 keys, most of the code is shared so we are using a single
+        # function with parameters.
+
+        # Prep placeholder variables for product_definition_section.
+        discipline = mock.sentinel.discipline
+        tablesVersion = mock.sentinel.tablesVersion
+        rt_coord = DimCoord(24, 'forecast_reference_time',
+                            units='hours since epoch')
+        metadata = empty_metadata()
+
+        # Use the section 4 from either product_definition_section #1 or #31.
+        # #1 contains fixed surface elements, #31 does not.
+        templates = {0: template_0(), 31: template_31()}
+        template_number = 0 if fs_is_expected else 31
+        section_4 = templates[template_number]
+        section_4.update({
+            'productDefinitionTemplateNumber': template_number,
+            'parameterCategory': None,
+            'parameterNumber': None
+        })
+
+        fixed_surface_keys = [
+            'typeOfFirstFixedSurface',
+            'scaledValueOfFirstFixedSurface',
+            'typeOfSecondFixedSurface'
+        ]
+        for key in fixed_surface_keys:
+            # Force the presence or absence of the fixed surface elements even
+            # when they're respectively ignored or expected.
+            if fs_is_present and key not in section_4:
+                section_4[key] = template_0()[key]
+            elif (not fs_is_present) and key in section_4:
+                del section_4[key]
+
+        def run_function():
+            # For re-use in every type of test below.
+            product_definition_section(
+                section_4, metadata, discipline, tablesVersion, rt_coord)
+
+        if fs_is_expected and not fs_is_present:
+            # Should error since the expected keys are missing.
+            with self.assertRaises(KeyError) as context:
+                run_function()
+            self.assertTrue('FixedSurface' in str(context.exception))
+        else:
+            # Should have a successful run for all other circumstances.
+            run_function()
+            # Translate_phenomenon_patch is the end of the function,
+            # and should be able to accept None for the fixed surface
+            # arguments. So should always have run.
+            self.assertEqual(self.translate_phenomenon_patch.call_count, 1)
+            phenom_call_args = self.translate_phenomenon_patch.call_args[1]
+            for key in fixed_surface_keys:
+                # Check whether None or actual values have been passed for
+                # the fixed surface arguments.
+                if fs_is_expected:
+                    self.assertEqual(phenom_call_args[key], section_4[key])
+                else:
+                    self.assertEqual(phenom_call_args[key], None)
```

Comment:
@trexfeathers You can use `self.assertIsNone(phenom_call_args[key])` instead, if you wish.

Reply:
👍